kythe/go/localrun/localrun.go

```diff
@@ -0,0 +1,503 @@
+	_, ok := ls[l]
+	return ok
+}
+
+func (ls LanguageSet) hasExtractor(e string) bool {
+	for l := range ls {
+		if l.kZipExtractorName() == e {
+			return true
+		}
+	}
+	return false
+}
+
+// LanguageMap is a mapping from the string name of each language to the enum
+// value.
+var LanguageMap map[string]Language = makeLanguageMap()
+
+func makeLanguageMap() map[string]Language {
+	r := map[string]Language{}
+	for l := range AllLanguages() {
+		r[strings.ToLower(l.String())] = l
+	}
+	return r
+}
+
+const (
+	// Ordered linear set of states the localrun program can be in.
+	nothing mode = iota
+	extract
+	index
+	postprocess
+	serve
+)
+
+type mode int
+
+func (m mode) String() string {
+	return [...]string{"nothing", "extract", "index", "postprocess", "serve"}[m]
+}
+
+// Runner is responsible for indexing the repo with Kythe.
+type Runner struct {
+	// Generally useful configuration
+	KytheRelease   string
+	WorkingDir     string
+	OutputDir      string
+	WorkerPoolSize int
+	CacheSize      *datasize.Size
+	graphStore     graphstore.Service
+
+	// Building/extracting options.
+	Languages LanguageSet
+	Targets   []string
+
+	// Serving configuration options.
+	Port int
+
+	// Timeout for indexing.
+	Timeout time.Duration
+
+	// Internal state
+	m          mode
+	indexedOut io.Reader
+	besFile    string
+}
+
+func (r *Runner) checkMode(m mode) error {
+	if r.m != m-1 {
+		return fmt.Errorf("couldn't transition to state %q from mode %q", m, r.m)
+	}
+
+	r.m = m
+	return nil
+}
+
+// Extract builds the provided targets and extracts compilation results.
+func (r *Runner) Extract(ctx context.Context) error {
+	if err := r.checkMode(extract); err != nil {
+		return err
+	}
+
+	tmpfile, err := ioutil.TempFile("", "bazel_build_event_protocol.pb")
+	if err != nil {
+		return fmt.Errorf("error creating tmpfile: %v", err)
+	}
+	r.besFile = tmpfile.Name()
+
+	// TODO: This is just a simple translation of the compile script into
+	args := append([]string{
+		fmt.Sprintf("--bazelrc=%s/extractors.bazelrc", r.KytheRelease),
+		"build",
+		"--keep_going",
+		// TODO: Is this a good idea? It makes it a lot easier to
+		// develop this since you can see successes/failures but may
+		// break CI workflows.
+		"--color=yes",
+		fmt.Sprintf("--override_repository=kythe_release=%s", r.KytheRelease),
+		fmt.Sprintf("--build_event_binary_file=%s", r.besFile),
+		"--",
+	},
+		r.Targets...)
+	fmt.Fprintf(os.Stderr, "Building with event file at: %s\n", r.besFile)
+
+	cmd := exec.CommandContext(ctx,
+		"bazel",
+		args...)
+	cmd.Stderr = os.Stderr
+	cmd.Dir = r.WorkingDir
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("error: %v", err)
+	}
+
+	return nil
+}
+
+// Index indexes the extracted data.
+// Index is a reimplementation of the https://github.com/kythe/kythe/blob/master/kythe/release/kythe.sh script's --index mode.
+func (r *Runner) Index(ctx context.Context) error {
+	if err := r.checkMode(index); err != nil {
+		return err
+	}
+
+	kzips := []string{}
```

Comment:
`var kzips []string` is the idiomatic declaration.

Reply:
Done